### PR TITLE
resolved_ts: add endpoint

### PR DESCRIPTION
**Pull request by 5kbpers**
Signed-off-by: 5kbpers <tangminghua@pingcap.com>

### What problem does this PR solve?

Problem Summary:
Add endpoint for resolved ts.

### What is changed and how it works?

What's Changed:
* Add endpoint
* Add `rts_id` to `apply::ObserveCmd` for temporarily being compatible with CDC
* Run resolved ts worker when starting TiKV server

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
* N/A

### CODE REVIEW

File: components/raftstore/src/store/fsm/store.rs [hunk 1]
```diff
@@ -1275,6 +1279,8 @@ impl<EK: KvEngine, ER: RaftEngine> RaftBatchSystem<EK, ER> {
                 let peer = peer_fsm.get_peer();
                 meta.readers
                     .insert(peer_fsm.region_id(), ReadDelegate::from_peer(peer));
+                meta.peer_properties
+                    .insert(peer_fsm.region_id(), Arc::default());
```
Comment (NingLin-P) [inline components/raftstore/src/store/fsm/store.rs:1283]: Should remove it when a region is destroyed

File: components/resolved_ts/src/endpoint.rs [hunk 2]
```diff
@@ -1,28 +1,459 @@
...
+                ScanEntry::TxnEntry(_) => panic!("unexpected entry type"),
+            }
+        }
+    }
+}
+
+pub struct Endpoint<T, E: KvEngine, C> {
+    store_meta: Arc<Mutex<StoreMeta>>,
+    regions: HashMap<u64, ObserveRegion>,
+    // raft_router: T,
```
Comment (NingLin-P) [inline components/resolved_ts/src/endpoint.rs:166]:
```suggestion
```

File: components/resolved_ts/src/lib.rs [hunk 3]
```diff
@@ -1,5 +1,14 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
+//! Resolved TS is a timestamp that represents the lower bonud of incoming Commit TS
+//  and the upper bound of outgoing Commit TS.
+//! Through this timestamp we can get a consistent view in the transaction level.
+//!
+//! To maintain a correct Resolved TS, these premises must be satisfied:
+//!   1. Tracing all locks in the region, use the minimal Start TS as Resolved TS.
+//!   2. If there is not any lock, use the latest timestamp as Resolved TS.
+//!   3. Resolved TS must be advanced by the region leader.
```
Comment (NingLin-P) [inline components/resolved_ts/src/lib.rs:10]:
```suggestion
//!   3. Resolved TS must be advanced by the region leader after it has applied on its term.
```

**Comment by NingLin-P** [comment]
Please resolve the conflict.

File: components/resolved_ts/src/endpoint.rs [hunk 4]
```diff
@@ -1,28 +1,459 @@
...
+                                key,
+                                start_ts,
+                                commit_ts,
+                                ..
+                            } => locks.push(PendingLock::Untrack {
+                                key: key.clone(),
+                                start_ts: *start_ts,
+                                commit_ts: *commit_ts,
+                            }),
+                            _ => (),
```
Comment (NingLin-P) [inline components/resolved_ts/src/endpoint.rs:102]: Why not handle `ChangeRow::OnePc` here?

File: components/resolved_ts/src/endpoint.rs [hunk 5]
```diff
@@ -1,28 +1,459 @@
...
+                    match log {
+                        ChangeLog::Error(e) => return Err(Error::Request(e.clone())),
+                        ChangeLog::Rows { rows, .. } => rows.iter().for_each(|row| match row {
+                            ChangeRow::Prewrite { key, start_ts, .. } => {
+                                self.resolver.track_lock(*start_ts, key.to_raw().unwrap())
+                            }
+                            ChangeRow::Commit { key, .. } => {
+                                self.resolver.untrack_lock(&key.to_raw().unwrap())
+                            }
+                            _ => (),
```
Comment (NingLin-P) [inline components/resolved_ts/src/endpoint.rs:118]: Ditto

File: components/raftstore/src/coprocessor/mod.rs [hunk 6]
```diff
@@ -232,9 +235,9 @@ impl CmdBatch {
 
 pub trait CmdObserver<E>: Coprocessor {
     /// Hook to call after preparing for applying write requests.
-    fn on_prepare_for_apply(&self, observe_id: ObserveID, region_id: u64);
+    fn on_prepare_for_apply(&self, cdc_id: ObserveID, rts_id: ObserveID, region_id: u64);
```
Comment (overvenus) [inline components/raftstore/src/coprocessor/mod.rs:238]:
```suggestion
    fn on_prepare_for_apply(&self, cdc_id: ObserveID, rts_id: ObserveID, region_id: u64);
```

Why do we need two ObserveID?

Comment (NingLin-P) [reply components/raftstore/src/coprocessor/mod.rs:238]: One for cdc worker, one for resolved ts worker, commands will send to both these two workers respectively.

**Review by NingLin-P — APPROVED** [review]

File: components/raftstore/src/store/fsm/apply.rs [hunk 7]
```diff
@@ -2852,19 +2854,51 @@ impl ObserveID {
 
 struct ObserveCmd {
     // TODO: Add flags to disable observing.
-    id: ObserveID,
+    cdc_id: ObserveID,
```
Comment (nolouch) [inline components/raftstore/src/store/fsm/apply.rs:2857]: Under what scenarios will `cdc_id` and `rst_id` be different?

File: src/config.rs [hunk 8]
```diff
@@ -2285,6 +2286,7 @@ impl Default for CdcConfig {
             min_ts_interval: ReadableDuration::secs(1),
             old_value_cache_size: 1024 * 1024,
             hibernate_regions_compatible: true,
+            scan_lock_pool_size: 2,
```
Comment (nolouch) [inline src/config.rs:2289]: What is the pool size in before?

File: components/resolved_ts/src/endpoint.rs [hunk 9]
```diff
@@ -98,3 +524,36 @@ impl<S: Snapshot> fmt::Display for Task<S> {
...
+            Task::ChangeLog {
+                cmd_batch,
+                snapshot,
+            } => self.handle_change_log(cmd_batch, snapshot),
+            Task::ScanLocks {
+                region_id,
+                observe_id,
+                entries,
+            } => self.handle_scan_locks(region_id, observe_id, entries),
+            Task::RegisterAdvanceEvent => self.register_advance_event(),
```
Comment (nolouch) [inline components/resolved_ts/src/endpoint.rs:556]: need with this #9869 change?

File: components/raftstore/src/store/fsm/apply.rs (hunk 7 above)
Comment (nolouch) [inline components/raftstore/src/store/fsm/apply.rs:2857]: Once we extract the  `resolve ts` completely, is the obeserver id of the cdc still needed? Can we keep only one ?

**Review by NingLin-P — COMMENTED** [review]
.

File: src/config.rs (hunk 8 above)
Comment (NingLin-P) [reply src/config.rs:2289]: https://github.com/tikv/tikv/blob/eb57a2aeaeb29a49fe9c8101c8b61d4128d27110/components/cdc/src/endpoint.rs#L284-L291

File: components/resolved_ts/src/endpoint.rs (hunk 9 above)
Comment (NingLin-P) [reply components/resolved_ts/src/endpoint.rs:556]: @5kbpers PTAL

File: components/raftstore/src/store/fsm/apply.rs (hunk 7 above)
Comment (NingLin-P) [reply components/raftstore/src/store/fsm/apply.rs:2857]: I think it is unneeded, the `cdc` worker and `resolved-ts` worker  will share the same observer /cc @5kbpers

**Review by JmPotato — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Comment by hi-rustin** [comment]
@5kbpers consistent failures. PTAL~

**Pull request merged**